flight/PiOS/Common/pios_ms5xxx.c

```diff
@@ -0,0 +1,698 @@
+		return -2;
+	if (dev->dev_id == 0)
+		return -3;
+	return 0;
+}
+
+
+/**
+ * @brief PIOS_MS5XXX_Init Initializes the MS5xxx chip
+ * @param cfg Configuration structure. Note that it is not deep copied.
+ * @return 0 on success. -1 if the allocation fails, -2 if the board
+ * setup file does not define the model of chip, and -3 if the write
+ * command fails.
+ */
+int32_t PIOS_MS5XXX_Init(const struct pios_ms5xxx_cfg *cfg)
+{
+	// Set the pointer to the configuration struct. Note that it is not deep copied.
+	dev->cfg = cfg;
+
+	// Check that a model of the chip was chosen
+	if (cfg->pios_ms5xxx_model == 0) {
+		return -2;
+	}
+
+	if (PIOS_MS5XXX_WriteCommand(MS5XXX_RESET) != 0) {
+		return -3;
+	}
+
+	PIOS_DELAY_WaitmS(20);
+
+	uint8_t data[2];
+
+	/* Calibration parameters */
+	for (int i = 0; i < NELEMENTS(dev->calibration); i++) {
+		PIOS_MS5XXX_Read(MS5XXX_CALIB_ADDR + i * 2, data, 2);
+		dev->calibration[i] = (data[0] << 8) | data[1];
+	}
+
+	PIOS_SENSORS_Register(PIOS_SENSOR_BARO, dev->queue);
+
+	dev->task = PIOS_Thread_Create(
+			PIOS_MS5XXX_Task, "pios_ms5xxx", MS5XXX_TASK_STACK_BYTES, NULL, MS5XXX_TASK_PRIORITY);
+	PIOS_Assert(dev->task != NULL);
+
+	return 0;
+}
+
+/**
+ * Initialise the MS5XXX sensor
+ */
+int32_t PIOS_MS5XXX_SPI_Init(uint32_t spi_bus_id, uint32_t slave_num, const struct pios_ms5xxx_cfg *ms5xxx_cfg)
+{
+	// Allocate the memory
+	dev = (struct ms5xxx_dev *)PIOS_MS5XXX_alloc();
+	if (dev == NULL) {
+		return -1;
+	}
+
+	// Do a couple SPI specific things
+	dev->pios_bus_type = PIOS_SPI_BUS;
+	dev->dev_id = spi_bus_id;
+	dev->slave_num = slave_num;
+
+	// Go on to the generic initialization
+	return PIOS_MS5XXX_Init(ms5xxx_cfg);
+}
+
+int32_t PIOS_MS5XXX_I2C_Init(int32_t i2c_bus_id, enum MS5XXX_I2C_ADDRESS i2c_address, const struct pios_ms5xxx_cfg *ms5xxx_cfg)
+{
+	// Allocate the memory
+	dev = (struct ms5xxx_dev *)PIOS_MS5XXX_alloc();
+	if (dev == NULL) {
+		return -1;
+	}
+
+	// Do a couple I2C specific things
+	dev->pios_bus_type = PIOS_I2C_BUS;
+	dev->dev_id = i2c_bus_id;
+	dev->ms5xxx_i2c_addr = i2c_address;
+
+	// Go on to the generic initialization
+	return PIOS_MS5XXX_Init(ms5xxx_cfg);
+}
+
+/**
+ * Claim the MS5XXX device semaphore.
+ * \return 0 if no error
+ * \return -1 if timeout before claiming semaphore
+ */
+static int32_t PIOS_MSXXX_ClaimDeviceSemaphore(void)
+{
+	PIOS_Assert(PIOS_MS5XXX_Validate(dev) == 0);
+
+	return PIOS_Semaphore_Take(dev->busy, PIOS_SEMAPHORE_TIMEOUT_MAX) == true ? 0 : 1;
+}
+
+/**
+ * Release the MS5XXX device semaphore.
+ * \return 0 if no error
+ */
+static int32_t PIOS_MS5XXX_ReleaseDeviceSemaphore(void)
+{
+	PIOS_Assert(PIOS_MS5XXX_Validate(dev) == 0);
+
+	return PIOS_Semaphore_Give(dev->busy) == true ? 0 : 1;
+}
+
+/**
+ * @brief Claim the SPI bus for the baro communications and select this chip
+ * @return 0 if successful, -1 for invalid device, -2 if unable to claim bus
+ */
+static int32_t PIOS_MS5XXX_ClaimSPIBus(void)
+{
+#if defined(PIOS_INCLUDE_SPI)
```

Comment:
These #ifdefs were included because of compile-time failures when `PIOS_INCLUDE_SPI` wasn't set. I will look into why the transfer byte and transfer block sections (~L497) don't cause compiler failures (because I kinda think they should).